### PR TITLE
Sketcher: Fix several issues in SketchObject::trim

### PR DESCRIPTION
**Pull request by davidosterberg**
This is a PR that addresses multiple issues in the SketchObject::trim method. 

**Issues addressed:**
- exception when trimming a loose end of a Arc of circle
- conflicting constraints when trimming
- loss of tangency when trimming arcs tangent to lines or other curves
- incorrect trimming when the intersecing geometry has a PointOnObject constraint to the arc, but also another intersection og geometry.

**Fixes:**
https://tracker.freecadweb.org/view.php?id=4066
https://tracker.freecadweb.org/view.php?id=3910

Also see forum discussion
https://forum.freecadweb.org/viewtopic.php?f=10&t=53299

**Selected demos**
Before PR (loss of tangency)
test4_before
After:
test4_after

Before PR (constraints modified wrongly)
test8_before
After PR
test8_after

Before PR (uncaught exception and loss of constraint)
test5_before
After PR
test5_after

Before PR (trim results in unsolvable sketch)
test6_before
After PR:
test6_after

### CODE REVIEW

**Comment by chennes** [comment]
Very nice, this will make trim much more useful. In your first example, I end up with redundant constraints: if I've already got the endpoints of the lines constrained to the circles, when the trim operation adds the coincident constraints to the arc endpoints those point-on-object constraints are now redundant: have you considered automatically removing them?

**Comment by davidosterberg** [comment]
> Very nice, this will make trim much more useful. 

Thanks I hope so. :)


> In your first example, I end up with redundant constraints: if I've already got the endpoints of the lines constrained to the circles, when the trim operation adds the coincident constraints to the arc endpoints those point-on-object constraints are now redundant: have you considered automatically removing them?

For me it works as I drew it (with end-to-curve tangency constraints).
But I can reproduce the redundant constraints error if I add the PointOnObject and then the simple Tangent separately. 
image
The code doesnt handle that case correctly. I will see if I can fix it.

**Comment by davidosterberg** [comment]
Alright, I have fixed the bug and also refactored a bit to avoid code repetition.

**Comment by abdullahtahiriyo** [comment]
Reviewing now (mutex).

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -2292,6 +2298,44 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
...
+                    constrType1 = Sketcher::Coincident;
+                    secondPos1 = constr->FirstPos;
+                    handlemultipleintersection(constr, GeoId, start, secondPos1);
+                    delete_list.push_back(constrId);
+                }
+            }
+            constrId++;
+        }
+        constrId=0;
+        for (std::vector<Constraint *>::const_iterator it=constraints.begin(); it != constraints.end(); ++it) {
```
Comment (abdullahtahiriyo) [inline src/Mod/Sketcher/App/SketchObject.cpp:2319]: It looks to me that this for loop may be merged with the for loop above (single loop with two if inside its body). Am I wrong?

**Comment by abdullahtahiriyo** [comment]
Notes to self:
- delConstraintOnPoint with false argument is only used for trimming and filleting. When trimming a line containing a vertical/horizontal alignment on an end point, it appears reasonable not to transfer the alignment constraint as a user can be expected not to want another geometry to follow the trimmed geometry. When filleting the point, it is expected that the user would like to continue alignment where the corner was, but the corner has moved, so sometimes not transferring would be the wanted behaviour and sometimes transferring would be the wanted behaviour. A default of not transferring appears reasonable.

- Magic number: 500*Precision::Confusion(), depends on the physical length of the sketch. Different if working on a road blueprint or a nano-machine. Will likely affect trimming in the second example. This magic number is used to correct for a deficiency either on seekTrimPoints or in the solver which does not achieve sufficient convergence. It fixes the behaviour (at least in a range of cases), but it does not fix the root cause.

**Review by abdullahtahiriyo — APPROVED** [review]
Thanks for your PR.

It would be great if you could take a look to the double for loop in the lambda function to consider if it could be possible to merge it into a single loop.

For the rest, the PR refactors some code. I have not tested every single combination. It appears to work reasonably well and we will get feedback from the community after merge.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 1 above)
Comment (davidosterberg) [reply src/Mod/Sketcher/App/SketchObject.cpp:2319]: I did this for a reason (initially I incuded everything in one single loop)

It is possible that the Sketch has both PointOnObject constrait to the intersecting line and a Simple Tangent. In this case we want to change to a single end-to-end tangency, i.e we want to ensure that constrType1 is set to Sketcher::Tangent.
The double loop ensures that the ordering of the constraints does not influence the result. 

Of course a double loop is more expensive than a single loop. But as this only happens when trimming and it is very quick. I don't think it is worth figuring out how to merge the loops.

Comment (abdullahtahiriyo) [reply src/Mod/Sketcher/App/SketchObject.cpp:2339]: I understand. Please document this design choice in the code in a comment. If you do not, somebody is going to see that code and refactor it into a single loop leading to unwanted behaviour.

Comment (davidosterberg) [reply src/Mod/Sketcher/App/SketchObject.cpp:2343]: Ok, done. I also rebased and squashed the commits into one nice commit.

**Comment by abdullahtahiriyo** [comment]
Thank you.

**Pull request merged**